Refactor parser.c for better testability

brawner opened the pull request:
This PR separates out the `parser.c` functionality into separate source files with their own headers so additional unit tests can be added to target specific functions. I still need to create the additional unit tests, but this current refactor passes the tests that currently exist, which has about 88% coverage.

Because this PR is so large, I've separated the changes into logical commits. Reviewing each commit individually might be the easiest way to scan through this PR as they illustrate the copy-paste changes I did. They could be their own PRs if that's desired by the reviewers.

Main commits
* https://github.com/ros2/rcl/pull/754/commits/8872020a45979b47f5404050c885b63b4924ad50 Shuffles functions into their own separate files, but doesn't change any actual logic
* https://github.com/ros2/rcl/pull/754/commits/70e0c99357acbadef891d6d4bdd85fa3553db09f Reorders `parser.c` to match `parser.h`
* https://github.com/ros2/rcl/pull/754/commits/bbb23d26ab31d3aa779e8ab005af56b1541705ad Creates a couple of macros and reduces some of the duplicate code when copying rcl_yaml_variant_t
* https://github.com/ros2/rcl/pull/754/commits/5635514f10bc740762cbf07f419f7171c840008d Creates some more macros for deduplicating the add_value_to_array type functions.

Addressing PR Feedback commits
* https://github.com/ros2/rcl/pull/754/commits/e741afa1eb25f3b92bd44758e387d09a46674f97 Removes a debug fprintf
* https://github.com/ros2/rcl/pull/754/commits/e617716edae2175d9f3c7020bc6ef6c7ba61165a Various PR feedback
* https://github.com/ros2/rcl/pull/754/commits/3ff206f6388a2e530a802193a32df7e9e16a363c Moves impl headers into `src` directory

brawner commented:
Testing `--packages-select rcl_yaml_param_parser`

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12006)](http://ci.ros2.org/job/ci_linux/12006/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7033)](http://ci.ros2.org/job/ci_linux-aarch64/7033/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=9784)](http://ci.ros2.org/job/ci_osx/9784/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=11942)](http://ci.ros2.org/job/ci_windows/11942/)

wjwwood commented:
@anup-pem do you have time to review this pull request since you originally wrote this?  It would help us make sure it doesn’t change the code in a way that you didn’t intend or that might cause apex issues in the future.

hidmic commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -0,0 +1,832 @@
+
+  rcl_variant_t * param_value = &(params_st->params[node_idx].parameter_values[parameter_idx]);
+
+  // param_value->string_value = rcutils_strdup(value, allocator);
```
@brawner why keep this?

brawner replied:
Removed

hidmic commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -0,0 +1,832 @@
+      ret = RCUTILS_RET_ERROR;
+      break;
+    case DATA_TYPE_BOOL:
+      if (false == is_seq) {
```
@brawner probably preexistent, but:
```suggestion
      if (!is_seq) {
```
? Same below.

brawner replied:
Definitely preexistent. Switched

hidmic commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -0,0 +1,832 @@
+            break;
+          }
+
+          memmove(param_name, parameter_ns, params_ns_len);
```
@brawner perhaps unrelated, we can use `memcpy` instead.

brawner replied:
Switched to memcpy

brawner commented on rcl_yaml_param_parser/src/add_to_arrays.c:
```diff
@@ -0,0 +1,118 @@
+        RCUTILS_SAFE_FWRITE_TO_STDERR("Error allocating mem\n"); \
+        return RCUTILS_RET_BAD_ALLOC; \
+      } \
+      memmove(val_array->values, tmp_arr, (val_array->size * sizeof(value_type))); \
```
Switched to memcpy

brawner commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12136)](http://ci.ros2.org/job/ci_linux/12136/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7129)](http://ci.ros2.org/job/ci_linux-aarch64/7129/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=9880)](http://ci.ros2.org/job/ci_osx/9880/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12037)](http://ci.ros2.org/job/ci_windows/12037/)

wjwwood commented on rcl_yaml_param_parser/QUALITY_DECLARATION.md:
```diff
@@ -22,6 +22,7 @@ The current version can be found in its [package.xml](package.xml), and its chan
 All symbols in the installed headers are considered part of the public API.
 
 All installed headers are in the `include` directory of the package, headers in any other folders are not installed and considered private.
+Headers in the `include/rcl_yaml_param_parser/impl` directory should also be considered private.
```
If they're private why do they need to be in `include`? This is C, not C++, there should be no need to expose a symbol but also consider it to be private, I think. The only reason to do that in C++ is for templates which use functions and inline implementations in headers. Since neither of those are the case for C, you should be able to just move these into `src`, right?

brawner replied:
Moved headers into `src` directory

brawner replied:
Err, looks like functions need to be labeled `RCL_YAML_PARAM_PARSER_PUBLIC` not LOCAL for it to work with the new unit tests.

wjwwood replied:
They can be made visible without being in the include folder.

brawner replied:
Oh yah totally. In response to this comment, I moved the headers to source and switched it to `LOCAL`, but the followup force push was just me switching them back to `PUBLIC`. Headers are still in `src`.

brawner commented:
This will need to be rebased after #765 and #775 are merged

hidmic commented on rcl_yaml_param_parser/src/parser.c:
```diff
@@ -696,74 +167,113 @@ void rcl_yaml_node_struct_fini(
-    allocator.deallocate(param_st->node_names[*node_idx], allocator.state);
-    return ret;
+    rcl_yaml_node_struct_fini(params_st);
+    return false;
```
@brawner one thing I noticed while testing something else is that this function is finalizing its output argument on error, invalidating the output argument without the caller knowing. I'll push a fix.

hidmic replied:
See https://github.com/ros2/rcl/pull/776.

brawner commented:
Rebased this onto master after recent PR merges. I updated the commits in the description at the top of this PR if the reviews would rather step through the changes that way. They make it easier to see what changes and what doesn't.

brawner commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12163)](http://ci.ros2.org/job/ci_linux/12163/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7154)](http://ci.ros2.org/job/ci_linux-aarch64/7154/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=9897)](http://ci.ros2.org/job/ci_osx/9897/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12053)](http://ci.ros2.org/job/ci_windows/12053/)

hidmic commented on rcl_yaml_param_parser/src/namespace.c:
```diff
@@ -0,0 +1,190 @@
+
+    tot_len = ns_len + sep_len + name_len + 1U;
+
+    cur_ns = allocator.reallocate(cur_ns, tot_len, allocator.state);
```
@brawner if `allocator.reallocate` fails, `cur_ns` will leak.

brawner replied:
Moving to https://github.com/ros2/rcl/pull/780

hidmic commented on rcl_yaml_param_parser/src/namespace.c:
```diff
@@ -0,0 +1,190 @@
+      return RCUTILS_RET_BAD_ALLOC;
+    }
+    memmove((cur_ns + ns_len), sep_str, sep_len);
+    memmove((cur_ns + ns_len + sep_len), name, name_len);
```
@brawner same about `memcpy` being cheaper.

brawner replied:
Done

hidmic commented on rcl_yaml_param_parser/src/namespace.c:
```diff
@@ -0,0 +1,190 @@
+      }
+      if (NULL != last_idx) {
+        tot_len = ((size_t)(last_idx - cur_ns) + 1U);
+        cur_ns = allocator.reallocate(cur_ns, tot_len, allocator.state);
```
@brawner same about `cur_ns` leaking on `allocator.reallocate` failure.

brawner replied:
https://github.com/ros2/rcl/pull/780

brawner commented on rcl_yaml_param_parser/src/add_to_arrays.c:
```diff
@@ -0,0 +1,119 @@
+      if (NULL == val_array->values) { \
+        val_array->values = tmp_arr; \
+        RCUTILS_SAFE_FWRITE_TO_STDERR("Error allocating mem\n"); \
+        return RCUTILS_RET_BAD_ALLOC; \
```
The callers already have to handle deallocation of value in several instances. The logic is complicated and this change will lead to larger changes, which I think are out of scope for just adding coverage to this package.

brawner commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12182)](http://ci.ros2.org/job/ci_linux/12182/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7166)](http://ci.ros2.org/job/ci_linux-aarch64/7166/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=9909)](http://ci.ros2.org/job/ci_osx/9909/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12065)](http://ci.ros2.org/job/ci_windows/12065/)

brawner commented:
Rebasing onto master after merging #780. Testing one last time just because it changes a lot of stuff

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12183)](http://ci.ros2.org/job/ci_linux/12183/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7167)](http://ci.ros2.org/job/ci_linux-aarch64/7167/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=9910)](http://ci.ros2.org/job/ci_osx/9910/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12066)](http://ci.ros2.org/job/ci_windows/12066/)

brawner commented:
All ci.ros2.org tests pass, and tests pass with rpr job. The failures in the Rpr job are unrelated to this PR and have already been addressed in PRs to rcl I believe.
